### PR TITLE
merge me_dev to dev

### PR DESCRIPTION
**Pull request by muwyse**
This PR includes:
- Addition of LCE Fill network in BedRock to fix protocol deadlock #1034 
- Implementation of BedRock Burst interfaces in LCE and IO CCE #1025 
- Minor fixes in LCE #1051
- ME documentation update #1052 
- minor cleanup for merges
- #1059 
- #1060

### CODE REVIEW

File: ci/accelerator.sh
Comment (dpetrisko) [inline ci/accelerator.sh:36]: I believe this doesn't test the 4x configs?

File: bp_me/src/v/network/bp_me_bedrock_wormhole_header_encode_lce_resp.sv
Comment (dpetrisko) [inline bp_me/src/v/network/bp_me_bedrock_wormhole_header_encode_lce_resp.sv:13]: Can we raise an issue to combine all of these msg_to_wh modules?

File: bp_top/src/v/bp_sacc_vdp.sv
Comment (dpetrisko) [inline bp_top/src/v/bp_sacc_vdp.sv:94]: Add assert that this only supports <=64b requests?

**Review by dpetrisko — CHANGES_REQUESTED** [review]
About 2/3 of the way through reviewing. Some initial comments, but overall looks great. Feel free to address or wait until full review

File: bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv
Comment (dpetrisko) [inline bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv:1]: how would this affect our FPGA builds that rely on stream -> AXIL converters?

File: bp_me/src/v/lce/bp_lce_fill.sv
Comment (dpetrisko) [inline bp_me/src/v/lce/bp_lce_fill.sv:138]: For critical path, this requires being input buffered. I'd add an assertion that fill_data_buffer_els_lp > 1 (2 to maintain full throughput)

Comment (dpetrisko) [reply bp_me/src/v/lce/bp_lce_fill.sv:138]: (same applies to lce_cmd

File: bp_me/src/v/lce/bp_lce.sv
Comment (dpetrisko) [inline bp_me/src/v/lce/bp_lce.sv:393]: This seems overly pessimistic. Aren't reqs dependent just on req_ready_lo since we're buffering it?

File: bp_me/src/v/lce/bp_lce.sv
Comment (dpetrisko) [inline bp_me/src/v/lce/bp_lce.sv:222]: ready_and_i is implicitly an early interface, and this is a late signal, so not ideal

File: bp_me/src/v/lce/bp_lce_cmd.sv
Comment (dpetrisko) [inline bp_me/src/v/lce/bp_lce_cmd.sv:564]: Don't think we need the ~critical_tag_sent_r here, because once we send, we move onto the next state?

File: bp_me/src/v/lce/bp_lce_fill.sv
Comment (dpetrisko) [inline bp_me/src/v/lce/bp_lce_fill.sv:21]: Seems like a lot of duplicated logic with bp_lce_cmd. Are there protocol implications to combining the modules and just prioritizing messages from fill in the state machine?

File: bp_me/src/v/lce/bp_lce_req.sv
Comment (dpetrisko) [inline bp_me/src/v/lce/bp_lce_req.sv:208]: Don't understand this comment. The interface is v->yumi, the ready_o is not exposed on this interface

File: bp_me/src/v/lce/bp_lce_req.sv
Comment (dpetrisko) [inline bp_me/src/v/lce/bp_lce_req.sv:312]: // TODO: For all supported caches, requests have a single data beat

File: bp_top/src/v/bp_tile.sv
Comment (dpetrisko) [inline bp_top/src/v/bp_tile.sv:196]: why not 0+:$bits(lce_req_header)?

File: bp_top/src/v/bp_tile.sv
Comment (dpetrisko) [inline bp_top/src/v/bp_tile.sv:169]: unused?

File: bp_top/src/v/bp_tile.sv
Comment (dpetrisko) [inline bp_top/src/v/bp_tile.sv:164]: unused?

File: bp_top/src/v/bp_tile.sv
Comment (dpetrisko) [reply bp_top/src/v/bp_tile.sv:196]: applies for all tiles

File: bp_me/src/v/lce/bp_lce_fill.sv
Comment (dpetrisko) [reply bp_me/src/v/lce/bp_lce_fill.sv:21]: Another thought: lce_fill messages are only used for data, so they can be considered a subset of lce_cmd. One way to handle this a little more simply would be to have in bp_lce.sv a static arbitration between LCE fill and LCE cmd, and send that message to the bp_lce_cmd.sv module. That way, there's only a single port into the bp_lce_cmd module, and we get rid of all the cache_req interface arbitration

**Review by dpetrisko — CHANGES_REQUESTED** [review]
reviewed all modules

File: bp_me/src/v/network/bp_me_bedrock_wormhole_header_encode_lce_resp.sv
Comment (muwyse) [reply bp_me/src/v/network/bp_me_bedrock_wormhole_header_encode_lce_resp.sv:13]: #1053

File: ci/accelerator.sh
Comment (muwyse) [reply ci/accelerator.sh:36]: I disabled after realizing the confusing accelerator config schemes. I think I fixed the configs now, so I will re-enable and test.

File: bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv
Comment (muwyse) [reply bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv:1]: this module and the matching _send module were replaced by bp_me_bedrock_mem_to_link, which is a bi-directional wh-burst converter. Any existing stream -> AXIL conversion should be changed to be burst -> AXIL.

File: bp_me/src/v/lce/bp_lce_cmd.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce_cmd.sv:564]: correct. Same in e_bedrock_cmd_data and e_bedrock_cmd_uc_data, actually.

File: bp_me/src/v/lce/bp_lce.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce.sv:222]: I suppose it depends on whether fill_lce_resp_header_v_lo is considered early or late. I would say it is fairly early since it is connected directly to the early valid signal from the fill header buffer inside bp_lce_fill.

File: bp_top/src/v/bp_tile.sv
Comment (muwyse) [reply bp_top/src/v/bp_tile.sv:169]: yep, good catch! leftover from prior iteration of refactoring the network code

File: bp_top/src/v/bp_tile.sv
Comment (muwyse) [reply bp_top/src/v/bp_tile.sv:164]: yep!

File: bp_top/src/v/bp_tile.sv
Comment (muwyse) [reply bp_top/src/v/bp_tile.sv:196]: the burst-wh converter takes a wormhole encoded header as input, not the plain BedRock burst header. It does this because each BedRock network requires a unique wh header encode, but the wh-burst converter itself only needs to know about some widths of fields in the headers, but not the details of the protocol header.

This is somewhat related to your question above about the wh header encoders and #1053 

For now, I think we leave as is and revisit with the open issue.

Comment (muwyse) [reply bp_top/src/v/bp_tile.sv:196]: The current setup is similar to what existed before

File: ci/accelerator.sh
Comment (muwyse) [reply ci/accelerator.sh:36]: tests seem to pass, re-enabled.

File: bp_top/src/v/bp_sacc_vdp.sv
Comment (muwyse) [reply bp_top/src/v/bp_sacc_vdp.sv:94]: done

File: bp_me/src/v/lce/bp_lce.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce.sv:393]: Yes, I think it can be simplified a little. I will refactor and update. My understanding is this signal basically passes through the cache and is connected to the cache packet ready_o signal used in the pipeline-cache handshake (for both I$ and D$).

Specifically, fill_ready_lo is not needed as fill is ready once reset goes low. Credits do not need to be considered here, they should be considered with the cache_req_yumi decision. Command ready needs to be considered because we want to block any cache operation until after the command module resets the cache's tag and stat memories.

File: bp_me/src/v/lce/bp_lce_req.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce_req.sv:203]: This is related to the cache_req_busy_o discussion above. I think ready_o can be reduced to ~is_reset and the sync done / mode lines. The other two lines should be part of the cache_req_yumi_o signal.

Credits are consumed with the cache request packet yumi, so they need to be considered there to apply backpressure on the cache request interface. Removing the credit check on the yumi is a bug, which I've encountered before, where it becomes possible to issue one more request than there are credits.

File: bp_me/src/v/lce/bp_lce_req.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce_req.sv:208]: The concern was cache_req_v_i being derived from this module's ready_o (through cache_req_busy_o), which would create a combinational loop. This possibility is eliminated by fixing the ready_o/busy_o/cache_req_yumi_o signals as discussed in other review comments.

File: bp_me/src/v/lce/bp_lce_fill.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce_fill.sv:21]: Yes, fill is basically a subset of the command module. I think they need to be separate FSMs, and therefore having them be separate modules is nice for code clarity.

I would be very hesitant in trying to command these two modules without some detailed system analysis to prove it wouldn't deadlock the coherence protocol. Multiplexing command and fill into one channel would violate the protocol deadlock rule that a lower priority network (command) can not stall a higher priority network (fill). We would have to guarantee that the command message would always be able to make forward progress, in all situations.

I agree there is some potentially undesirable/unoptimal arbitration for the cache fill interface and response network, but with static arbitration that favors the fill module, fill will never get blocked by command.

File: bp_me/src/v/lce/bp_lce.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce.sv:222]: Another option is to add a 2-fifo on the outbound response header to reduce the path lengths, with the obvious cost of storage for two message headers (only header because response data only comes from the command module).

File: bp_top/src/v/bp_tile.sv
Comment (dpetrisko) [reply bp_top/src/v/bp_tile.sv:189]: oh this is a super misleading name then. Should be wh_hdr, not pr_hdr

Comment (dpetrisko) [reply bp_top/src/v/bp_tile.sv:189]: Fine with keeping open issues for functionality, but we should rename

**Review by dpetrisko — CHANGES_REQUESTED** [review]

File: bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv
Comment (dpetrisko) [reply bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv:1]: We either need to maintain the stream converters or write a multicore->stream wrapper because we can't break zynq-parrot in dev

File: bp_me/src/v/lce/bp_lce_fill.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce_fill.sv:136]: data buffer required for fill and command; assertions added.

File: bp_me/src/v/lce/bp_lce.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce.sv]: Okay, I have adjusted some signals. busy_o is raised on timeout or if request module is not ready. Request module only transitions out of its reset state when cache initialization is finished by Command module (now called cache_init_done_*). A single credit is consumed per LCE request sent, and the credit check is performed when sending the LCE request header. Request module is ready if not in reset and in a valid mode (uncached or cached with sync complete).

The credit counting is now consistent with the UCE in that every request requires a single credit (a request may be multiple flits; currently that is 1 or 2 flits).

File: bp_me/src/v/lce/bp_lce_req.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce_req.sv]: See thread above. Credits now consumed on request send, but only one credit per request (not one per flit).

File: bp_me/src/v/lce/bp_lce.sv
Comment (muwyse) [reply bp_me/src/v/lce/bp_lce.sv]: If the coh_noc_max_credits_p is meant to be a max flits in the network, then both UCE and LCE need further updating, and this will add complexity since it would be unwise to partially send a request

File: bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv
Comment (muwyse) [reply bp_me/src/v/network/bp_me_cce_to_mem_link_recv.sv:1]: Got it. Reverting the deletion of these files and related to keep zynq-parrot functional. I will add an issue for updating bp_axi_top so that we can remove and delete these modules.

File: bp_top/src/v/bp_tile.sv
Comment (muwyse) [reply bp_top/src/v/bp_tile.sv:189]: renamed

**Review by dpetrisko — APPROVED** [review]
Awesome work! we should run linux through this on FPGA before merging, but code is good to go

**Pull request merged**
